keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+**How Resource classes can solve this:**<br/>
+Vendors can use DevicePlugin API to propagate new hardware features, and provide best-practice ResourceClass spec to consume their new hardware or new hardware features on Kubernetes. Vendors don’t need to worry supporting this new hardware would break existing use cases on old hardware because the Kubernetes scheduler takes the resource metadata into account during pod scheduling, and so only pods that explicitly request this new hardware through the corresponding ResourceClass name will be allocated with such resources.
+
+## Objectives
+Essentially, a ResourceClass object maps a non-native compute resource with a specific set of properties to a portable name. Cluster admins can create different ResourceClass objects with the same generic name on different clusters to match the underlying hardware configuration. Users can then use the portable names to consume the matching compute resources. Through this extra abstraction layer, we are hoping to achieve the following goals:
+- **Allows workloads to request compute resources with wide range of properties in simple and standard way.** We propose to introduce a new `ComputeResource` API and a field, `ComputeResources` in the `Node.Status` to store a list of `ComputeResource` objects. Kubelet can use a `ComputeResource` object to encapsulate the resource metadata information associated with its underlying physical compute resources and propagate this information to the scheduler by appending it to the `ComputeResources` list in the `Node.Status`. With the resource metadata information, the Kubernetes scheduler can determine the fitness of a node for a container resource request expressed through ResourceClass name by evaluating whether the node has enough unallocated units of a ComputeResource matching the property constraints specified in the ResourceClass. This allows the Kubernetes scheduler to take resource property into account to schedule pods on the right node whose hardware configuration meets the specific resource metadata constraints.
+- **Allows cluster admins to configure and manage different kinds of non-native compute resources in flexible and simple ways.** A cluster admin creates a ResourceClass object that specifies a portable ResourceClass name (e.g., `fast-nic-gold`), and list of property matching constraints (e.g., `resourceName in (solarflare.com/fast-nic, intel.com/fast-nic)`, or `type=XtremeScale-8000`, or `bandwidth=100G`, or `zone in (us-west1-b, us-west1-c)`). The property matching constraints follow the generic LabelSelector format, which allows us to cover a wide range of resource specific properties. The cluster admin can then define and manage resource quota with the created ResourceClass object.
```

Comment:
@ConnorDoyle would not node status be sufficient for users to see what all compute resources are available, along with their properties for getting idea about what resources can be selected by a resource class. Also there will be 
```
type ResourceClassStatus struct {
        Allocatable int32
        Requested   int32
}
```
through which user will be able to see at any moment what is the consumption state. Any other suggestions?

Reply:
@ConnorDoyle were you refering this?:

>
>Non Goals
>Extends the current resource requirement API of the container spec. The current resource requirement API is <br/>basically a “name:value” list. A commonly arising question is whether we should extend this API to support <br/>resource metadata requirements. We can consider this as a possible extension orthogonal to the <br/>ResourceClass proposal. A primary reason we propose to introduce ResourceClass API first is because non-native compute resources usually lack standard resource properties. Although there are benefits to allow users to <br/>directly express their resource metadata requirements in their container spec, it may also compromise <br/>workload portability if not used carefully. By introducing ResourceClass as an additional resource abstraction<br/> layer, users can express their special resource requirements through a high-level portable name, and cluster admins can configure compute resources properly on different environments to meet such requirements. We feel this helps promote portability and separation of concerns, while still maintains API compatibility.
```